WTG.Analyzers/Analyzers/ToString/ToStringAnalyzer.cs

```diff
@@ -0,0 +1,119 @@
+		{
+			var cache = new FileDetailCache();
+
+			context.RegisterSyntaxNodeAction(
+				c => Analyze(c, cache),
+				SyntaxKind.InvocationExpression);
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
+			{
+				return;
+			}
+
+			var invoke = (InvocationExpressionSyntax)context.Node;
+
+			switch (invoke.Expression.Kind())
+			{
+				case SyntaxKind.SimpleMemberAccessExpression:
+					{
+						var member = (MemberAccessExpressionSyntax)invoke.Expression;
+
+						if (member.Name.Identifier.Text == WellKnownMemberNames.ObjectToString)
+						{
+							var symbol = (IMethodSymbol)context.SemanticModel.GetSymbolInfo(invoke, context.CancellationToken).Symbol;
+
+							if (symbol != null)
+							{
+								switch (symbol.ReceiverType.SpecialType)
+								{
+									case SpecialType.System_String:
+										context.ReportDiagnostic(Rules.CreateDontCallToStringOnAStringDiagnostic(InvokeLocation(invoke, member.OperatorToken)));
+										break;
+
+									case SpecialType.System_Enum:
+										if (symbol.Parameters.Length == 0 &&
+											IsEnumLiteral(member.Expression, context.SemanticModel, context.CancellationToken))
+										{
+											context.ReportDiagnostic(Rules.CreatePreferNameofOverCallingToStringOnAnEnumDiagnostic(invoke.GetLocation()));
+										}
+										break;
+								}
+							}
+						}
+					}
+					break;
+
+				case SyntaxKind.MemberBindingExpression:
+					{
+						var binding = (MemberBindingExpressionSyntax)invoke.Expression;
+
+						if (binding.Name.Identifier.Text == WellKnownMemberNames.ObjectToString)
+						{
+							var symbol = (IMethodSymbol)context.SemanticModel.GetSymbolInfo(invoke, context.CancellationToken).Symbol;
+
+							if (symbol != null && symbol.ReceiverType.SpecialType == SpecialType.System_String)
+							{
+								context.ReportDiagnostic(Rules.CreateDontCallToStringOnAStringDiagnostic(InvokeLocation(invoke, binding.OperatorToken)));
+							}
+						}
+					}
+					break;
+			}
+		}
+
+		static bool IsEnumLiteral(ExpressionSyntax expression, SemanticModel model, CancellationToken cancellationToken)
+		{
+			while (expression.IsKind(SyntaxKind.ParenthesizedExpression))
+			{
+				var paren = (ParenthesizedExpressionSyntax)expression;
+				expression = paren.Expression;
+			}
+
+			var memberSymbol = model.GetSymbolInfo(expression, cancellationToken).Symbol;
+
+			return memberSymbol != null &&
+				memberSymbol.Kind == SymbolKind.Field &&
+				memberSymbol.IsStatic &&
+				memberSymbol.ContainingType.EnumUnderlyingType != null;
+		}
+
+		static Location InvokeLocation(InvocationExpressionSyntax invoke, SyntaxToken operatorToken)
+		{
+			//   -->...........<--
+			// value.ToString()
+			return Location.Create(
+				invoke.SyntaxTree,
+				TextSpan.FromBounds(
+					operatorToken.Span.Start,
+					invoke.ArgumentList.CloseParenToken.Span.End));
```

Comment:
What happens to incomplete code, e.g. `value.ToString(`?

Reply:
It seems to use the preceding token. I'll switch to `invoke.Span.End`, it gives the same result but is clearer and more direct. I'll also add a test to cover this case.